abmarl/sim/gridworld/state.py

```diff
@@ -23,58 +25,358 @@ class PositionState(StateBaseComponent):
+
+    @property
+    def no_overlap_at_rest(self):
+        """
+        Attempt to place each agent on its own cell.
+
+        Agents with initial positions will override this property.
+        """
+        return self._no_overlap_at_reset
+
+    @no_overlap_at_rest.setter
+    def no_overlap_at_reset(self, value):
+        assert type(value) is bool, "No overlap at reset must be a boolean."
+        self._no_overlap_at_reset = value
+
     def reset(self, **kwargs):
         """
         Give agents their starting positions.
 
         We use the agent's initial position if it exists. Otherwise, we randomly
         place the agents in the grid.
         """
+        self.grid.reset()
+
+        # Build lists of available positions
+        self._build_available_positions()
+
+        # Place agents with initial positions.
+        for agent in self.agents.values():
+            if agent.initial_position is not None:
+                self._place_initial_position_agent(agent)
+
+        # Now place agents with variable positions
+        for agent in self.agents.values():
+            if agent.initial_position is None:
+                self._place_variable_position_agent(agent)
+
+    def _build_available_positions(self, **kwargs):
+        """
+        Define the positions that are available per encoding.
+        """
         max_encoding = max([agent.encoding for agent in self.agents.values()])
-        ravelled_available_positions = {
-            encoding: set(i for i in range(self.rows * self.cols))
+        self.ravelled_positions_available = {
+            encoding: [i for i in range(self.rows * self.cols)]
             for encoding in range(1, max_encoding + 1)
         }
+
+    def _update_available_positions(self, agent_just_placed, **kwargs):
+        """
+        Update the available positions based on the agent that was just placed.
+        """
+        for encoding, positions_available in self.ravelled_positions_available.items():
+            # Remove this cell from any encoding where overlapping is False
+            if self.no_overlap_at_reset or \
+                    encoding not in self.grid.overlapping.get(agent_just_placed.encoding, {}):
+                try:
+                    positions_available.remove(
+                        np.ravel_multi_index(agent_just_placed.position, (self.rows, self.cols))
+                    )
+                except ValueError:
+                    # Catch a value error because this cell might have already
+                    # been removed from this encoding
+                    continue
+
+    def _place_initial_position_agent(self, ip_agent_to_place, **kwargs):
+        """
+        Place an agent with an initial position.
+        """
+        ravelled_initial_position = np.ravel_multi_index(
+            ip_agent_to_place.initial_position,
+            (self.rows, self.cols)
+        )
+        assert ravelled_initial_position in \
```

Comment:
Don't check this assertion. For the initial position agents, we'll just rely on the logic in the grid itself. Place the message after the grid placement assertion.